### DISCUSSION TITLE
Сложение массивов без учёта предыдущего сложения Combiner.push

### DISCUSSION THREAD

**Discussion by Bondian** [Chimildic/goofy#135] [Принцип работы]
Имеем:
```
  let firstArray = [1, 2, 3];
  let secondeArray = [4, 5, 6];
  let thirdArray = [7, 8, 9];
  let firstPush = Combiner.push(firstArray, secondeArray);
  let secondPush = Combiner.push(firstArray, thirdArray);

  // получаемый результат firstPush 1, 2, 3, 4, 5, 6, 7, 8, 9     Странно... не понимаю, почему сюда попадает 7, 8, 9...
  // получаемый результат secondPush 1, 2, 3, 4, 5, 6, 7, 8, 9
```
Хочется получить с минимальными изменениями в коде такие результаты:
```
  // желаемый результат firstPush 1, 2, 3, 4, 5, 6
  // желаемый результат secondPush 1, 2, 3, 7, 8, 9
```

**Comment by Chimildic** [comment]
`Combiner.push` не создает новый массив. Он добавляет новые элементы к первому массиву из входного списка. Возвращает его же, просто для удобства в некоторых ситуациях. Поэтому `firstArray` содержит все, что в него добавляется на всем протяжении кода.

Наиболее простой способ избежать этого, первым параметром дать пустой массив `[]` и принять результат.
```js
let firstPush = Combiner.push([], firstArray, secondeArray);
let secondPush = Combiner.push([], firstArray, thirdArray);
```

Или вместо оригинального массива передать копию.
```js
let firstPush = Combiner.push(Selector.sliceCopy(firstArray), secondeArray);
let secondPush = Combiner.push(Selector.sliceCopy(firstArray), thirdArray);
```

**Reply by Bondian** [reply]
Спасибище! Мне идеально подходит первый способ с пустым массивом. Пошёл исправлять свои "гирлянды" из кода ))

**Comment by efremych** [comment]
Не понял, почему в firstPush 7,8,9? Не имеет значение, что firstArray изменялся позднее?

selector.keeprandom (arr1, 3);
Combiner.push (arr1, arr2);

у меня выбирается 3 рандомных трека из массива arr1+arr2?

**Reply by Chimildic** [reply]
Можно перефразировать, не понимаю вопрос

**Reply by efremych** [reply]
// получаемый результат firstPush 1, 2, 3, 4, 5, 6, 7, 8, 9     Странно... не понимаю, почему сюда попадает 7, 8, 9...

Тут реально есть 7,8,9?

 firstArray содержит все, что в него добавляется на всем протяжении кода.

если после firstPush сделать console.log (firstArray), то будет выведено 1,2,3,4,5,6,7,8,9? Не имеет значение, что 7,8,9 будет добавлено позднее согласно коду?

**Reply by Chimildic** [reply]
Конечно имеет значение. Просто в примере _опрос результата_ происходит после всех телодвижений. 
Если отработать до второго _push_, то не будет _7,8,9_.

**Reply by efremych** [reply]
В примере опрашивают firstPush в конце, но оно же зафиксировалось на 1,2,3,4,5,6. Или до самого конца кода firstPush зависит от конечного вида firstArray?
Надо завтра самому попробовать на ПК, а то чувствую, усложняется картина мира)

**Reply by Chimildic** [reply]
`Combiner.push` не создает новый массив. `firstPush` не является новым массивом. Это **ссылка** с новым именем на `firstArray`. Поэтому любое изменение `firstArray` отражается на остальных. 
При `sliceCopy` или пустом массиве `[]` наоборот создается новый и не имеет отношения к `firstArray`.

**Reply by efremych** [reply]
Ага, кажется, удалось сложить пазл в голове…) спасибо большое!
